cloud-controller-manager/do/loadbalancers.go

```diff
@@ -275,6 +274,56 @@ func (l *loadBalancers) EnsureLoadBalancer(ctx context.Context, clusterName stri
 	}, nil
 }
 
+func (l *loadBalancers) updateLoadBalancer(ctx context.Context, lb *godo.LoadBalancer, lbRequest *godo.LoadBalancerRequest, service *v1.Service) (*godo.LoadBalancer, error) {
+	var lbCertID string
+	for _, rule := range lb.ForwardingRules {
+		if rule.TlsPassthrough {
+			continue
+		}
+		if rule.CertificateID != "" {
+			lbCertID = rule.CertificateID
+			// it's not clear that it is even possible for different forwarding rules
+			// on the same loadbalancer to have different certificates associated
+			// with them -- as far as we are concerned where a load balancer is
+			// created by CCM in the first place it doesn't seem possible so we only
+			// need to grab the first one we find
+			break
+		}
+	}
+
+	checkServiceCertID := true
+	if lbCertID != "" && lbCertID != getCertificateID(service) {
+		lbCert, _, err := l.resources.gclient.Certificates.Get(ctx, lbCertID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get current certificate for lb")
+		}
+
+		if lbCert.Type == certTypeLetsEncrypt {
+			l.ensureCertificateIDAnnot(service, lbCertID)
+			checkServiceCertID = false
+		}
+	}
+
+	if checkServiceCertID {
+		_, _, err := l.resources.gclient.Certificates.Get(ctx, getCertificateID(service))
+		if err != nil {
+			respErr, ok := err.(*godo.ErrorResponse)
+			if ok && respErr.Response.StatusCode == http.StatusNotFound {
+				fmtStr := "certificate UUID %q not found; the %q service annotation refers to nonexistent DO Certificate"
+				respErr.Message = fmt.Sprintf(fmtStr, getCertificateID(service), annDOCertificateID)
+			}
+			return nil, err
+		}
+	}
```

Comment:
I think the logic around certificate validation should be moved into a dedicated function. AFAICS, that would also allow us to replace the `checkServiceCertID` variable / check by a premature return.

Reply:
My inclination is to factor out functions in the following situations:

* there is an immediate need to reuse a snippet of code
* there will obivously be need to reuse a snippet of code
* the function where the snippet currently resides is getting too unwieldy

I don't think any of those criteria apply in this situation, but I'll take a look once I've had an opportunity to add some test cases to validate what is here so far.

Reply:
Code fluency and natural scoping of functionality are additional concerns to me. Right now, `updateLoadBalancer` consists of "do stuff around certificates" and the actual update. The former isn't obvious on first sight, especially since it's not core business to updating.

I also think the code would be easier to understand as we could get rid of `checkServiceCertID`: No need to set and check conditionals when a `return nil` has the same effect.